cloudferrylib/utils/extensions.py

```diff
@@ -0,0 +1,80 @@
+# Copyright (c) 2016 Mirantis Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the License);
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an AS IS BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+# implied.
+# See the License for the specific language governing permissions and#
+# limitations under the License.
+
+import importlib
+import inspect
+import logging
+import os
+import pkgutil
+import threading
+
+LOG = logging.getLogger(__name__)
+
+
+_extensions = {}
+_extensions_lock = threading.Lock()
+
+
+def is_py(root_path, module_name):
+    filepath = os.path.join(root_path, *module_name.split('.')) + '.py'
+    return os.path.isfile(filepath)
+
+
+def available_extensions(base_class, path=None):
+    """
+    :rtype: list
+    :returns: list of cinder plugins
+    :raises: InvalidCinderPluginError if no plugins found
+    """
+
+    with _extensions_lock:
+        if base_class in _extensions:
+            return _extensions[base_class]
+
+        def print_log(m):
+            LOG.warning("Cannot import module '%s'", m)
+
+        extensions = []
+        LOG.debug("Scanning available '%s' extensions for %s",
+                  base_class, path or 'all folders')
+        if isinstance(path, basestring):
+            path = [path]
+        else:
+            path = list(path)
```

Comment:
Looks like even if `path` is already list, you'd still make it list again, like `[[path]]`